### ISSUE TITLE
Incompatible feature sizes

### ISSUE THREAD

**Issue by JoelRunevic** [yang-ze-kang/SCMIL#2]
Hello, thanks for making the code accessible!

I am trying to use SCMIL myself with the following configuration:

```json
{
    "hidden_size": 384,
    "deep": 1,
    "n_cluster": null,
    "cluster_size": 64,
    "feature_weight": 0,
    "as_backbone": false,
    "dropout_rate": 0.25,
    "with_softfilter": true,
    "use_filter_branch": true,
    "with_cssa": true,
    "return_IS": false,
    "return_patch_label": false
 }
```

The input dim for my features is 1024, and I am using the same number of heads as in the given code (8). In the forward pass of the ClusterLocalAttention class, I get an incompatability in the dimensions. Namely, in the line:

```python
qs,ks,vs = self.to_qkv(x).chunk(3, dim = -1)
```

x has the same dimensionality as the input_dim (1024), but the code assumes that the dimensionality is hidden_size*3. I am not sure if I have misread the code. 

Here is the code that I am using (adapted to fit into my own codebase nicer). Thank you very much in advance!

```python
import torch
import pdb
import math
import numpy as np
import torch.nn.functional as F
from torch import nn, einsum
from cuml import KMeans
from einops import rearrange, reduce
from .model_components.SoftFilter import SoftFilterLayer
from .model_components.AMIL import AMILLayer
from .configs_model import SCMILConfig
        
class SelfAttention(nn.Module):
    def __init__(self, dim, num_heads, head_size, dropout_prob=0.25):
        super(SelfAttention, self).__init__()
        innder_dim = num_heads*head_size
        self.num_heads = num_heads
        self.head_size = head_size

        self.to_qkv = nn.Linear(dim, innder_dim*3)
        self.attn_dropout = nn.Dropout(dropout_prob)
        self.to_out = nn.Sequential(
            nn.Linear(innder_dim, dim),
            nn.Dropout(dropout_prob)
        )

    def forward(self, x):
        q,k,v = self.to_qkv(x).chunk(3, dim = -1)
        q, k, v = map(lambda t: rearrange(t, 'b n (h d) -> b h n d', h = self.num_heads), (q, k, v))
        attention_scores = torch.matmul(q, k.transpose(-1, -2))
        attention_scores = attention_scores / math.sqrt(self.head_size)
        attention_probs = nn.Softmax(dim=-1)(attention_scores)
        attention_probs = self.attn_dropout(attention_probs)
        out = torch.matmul(attention_probs, v)
        out = rearrange(out, 'b h n d -> b n (h d)', h = self.num_heads)
        out = self.to_out(out)
        return out


class ClusterLocalAttention(nn.Module):
    
    def __init__(self,hidden_size=384,n_cluster=None,cluster_size=None,feature_weight=0,dropout_rate=0.1) -> None:
        super().__init__()
        self.hidden_size = hidden_size
        self.n_cluster = n_cluster
        self.cluster_size = cluster_size
        self.feature_weight = feature_weight
        self.num_heads = 8

        self.atten = SelfAttention(dim=hidden_size, num_heads=self.num_heads,head_size=hidden_size//8)
        self.head_size = hidden_size//self.num_heads
        innder_dim = self.num_heads*self.head_size
        self.to_qkv = nn.Linear(hidden_size, innder_dim*3)
        self.attn_dropout = nn.Dropout(dropout_rate)
        self.to_out = nn.Sequential(
            nn.Linear(innder_dim, hidden_size),
            nn.Dropout(dropout_rate)
        )
    
    def forward(self,x,coords=None,cluster_label=None,return_patch_label=False):
        B,L,C = x.shape
        if cluster_label is not None:
            labels = cluster_label.numpy()[0]
        else:
            if self.n_cluster==None and self.cluster_size is not None:
                n_cluster = L//self.cluster_size
                if n_cluster==0:
                    n_cluster=1
            elif self.n_cluster is not None:
                n_cluster = self.n_cluster
            else:
                raise NotImplementedError
            if self.feature_weight==-1:
                np.random.seed(0)
                labels = np.random.randint(0,n_cluster,size=L)
            else:
                with torch.no_grad():
                    coords = coords.squeeze()
                    kmeans = KMeans(n_clusters=n_cluster, init='k-means++', tol=1e-4, max_iter=5, random_state=1)
                    coords = (coords-torch.mean(coords))/torch.std(coords)
                    if self.feature_weight!=0:
                        feats = x.squeeze()
                        feats = (feats-torch.mean(feats))/torch.std(feats)
                        feats = feats*self.feature_weight
                        coords = coords*(1-self.feature_weight)
                        coords = torch.cat([feats,coords],dim=1)
                    labels = kmeans.fit_predict(coords)
                    labels = labels.get()
        index = np.argsort(labels,kind='stable').tolist()
        index_ori = np.argsort(index,kind='stable').tolist()
        x = x[:,index]
        window_sizes = np.bincount(labels).tolist()
        # Prevent large cluster size after clustering
        window_sizes_new = []
        for size in window_sizes:
            if size>=self.cluster_size*2:
                num_splits = size // self.cluster_size
                quotient = size // num_splits
                remainder = size % num_splits
                result = [quotient + 1 if i < remainder else quotient for i in range(num_splits)]
                window_sizes_new.extend(result)
            else:
                window_sizes_new.append(size)
        window_sizes = window_sizes_new
        now = 0
        qs,ks,vs = self.to_qkv(x).chunk(3, dim = -1)
        qs, ks, vs = map(lambda t: rearrange(t, 'b n (h d) -> b h n d', h = self.num_heads), (qs, ks, vs))
        h = torch.zeros_like(qs)
        for i in range(len(window_sizes)):
            q,k,v = qs[:,:,now:now+window_sizes[i]],ks[:,:,now:now+window_sizes[i]],vs[:,:,now:now+window_sizes[i]]
            attention_scores = torch.matmul(q, k.transpose(-1, -2))
            attention_scores = attention_scores / math.sqrt(self.head_size)
            attention_probs = nn.Softmax(dim=-1)(attention_scores)
            attention_probs = self.attn_dropout(attention_probs)
            out = torch.matmul(attention_probs, v)
            h[:,:,now:now+window_sizes[i]] = out
            now += window_sizes[i]
        h = rearrange(h, 'b h n d -> b n (h d)', h = self.num_heads)
        h = self.to_out(h)+x
        if return_patch_label:
            return h,labels
        else:
            return h,None
        

class SCMIL(nn.Module):
    def __init__(self, config: SCMILConfig):
        super(SCMIL, self).__init__()
        self.n_classes = config.n_classes
        self.deep = config.deep
        self.n_cluster = config.n_cluster
        self.cluster_size = config.cluster_size
        self.feature_weight = config.feature_weight
        self.as_backbone = config.as_backbone
        self.with_softfilter = config.with_softfilter
        self.with_cssa = config.with_cssa
        self.input_size = config.in_dim
        self.dropout_rate = config.dropout_rate
        self.hidden_size = config.embed_dim
        self.return_IS = config.return_IS
        self.return_patch_label = config.return_patch_label
        self.coords = config.coords
        self.cluster_label = config.cluster_label


        if self.with_softfilter:
            config.use_filter_branch = True
        self.use_filter_branch = config.use_filter_branch
        
        if self.with_softfilter:
            self.softfilter = SoftFilterLayer(dim=self.input_size,hidden_size=self.hidden_size,deep=self.deep)
        
        if self.hidden_size==None:
            self.hidden_size = self.input_size
        self.fc1 = nn.Sequential(
            nn.Linear(self.input_size,self.hidden_size),
            nn.GELU(),
            nn.Dropout(self.dropout_rate )
        )
        if self.with_cssa:
            layers = []
            for _ in range(self.deep):
                layers.append(ClusterLocalAttention(hidden_size=self.hidden_size,n_cluster=self.n_cluster,cluster_size=self.cluster_size,feature_weight=self.feature_weight,dropout_rate=self.dropout_rate))
            self.attens = nn.Sequential(*layers)
        self.amil = AMILLayer(self.hidden_size, 256, dropout=self.dropout_rate)
        if not self.as_backbone:
            self.classifier = nn.Linear(self.hidden_size,self.n_classes)
        self.iter = 0

    def forward_attention(self, x, attn_only=False, return_intermeds=False):
        self.iter+=1
        res = {}
        if self.with_softfilter:
            x, logits = self.softfilter(x)
            if self.return_IS:
                res['IS'] = logits.squeeze().cpu().detach().numpy()
        if self.use_filter_branch:
            idx = torch.where(logits.squeeze()>0.5)[0]
            h = x[:,idx]
            self.coords = self.coords[:,idx]
            idx2 = torch.where(logits.squeeze()<=0.5)[0]
            h_app = x[:,idx2]
        else:
            h = self.fc1(x)
        if self.with_cssa:
            if h.shape[1]>1:
                for atten in self.attens:
                    h, patch_label = atten(h, self.coords, self.cluster_label, self.return_patch_label)
        if self.return_patch_label:
            patch_label_all = np.zeros(len(logits.squeeze()))-1
            patch_label_all[idx.cpu().numpy()] = patch_label
            res['patch_label'] = patch_label_all
        if self.return_IS or self.return_patch_label:
            return res
        if self.use_filter_branch:
            h = torch.cat([h,h_app],dim=1)
        h, att ,_ =self.amil(h)
        if self.as_backbone:
            return h
        return h, att
        
    def forward_no_loss(self, h, attn_mask=None, logits_only=True, return_intermeds=False, **kwargs):
        h, A = self.forward_attention(h, return_intermeds=return_intermeds)
        log_dict = {} # empty dict to avoid downstream errors

        A_raw = A 
        M = h 
        logits = self.classifier(M)

        if logits_only:
            return logits, log_dict 
        
        return logits, A_raw, A, M, log_dict 


    def forward(self, h, attn_mask=None, label=None, loss_fn=None, return_intermeds=False, logits_only=True, **kwargs):
        self.coords = kwargs.get('coords').float()

        out = self.forward_no_loss(h, attn_mask=attn_mask, logits_only=logits_only, **kwargs)
        if logits_only:
            logits, log_dict = out
        else:
            logits, A_raw, A, M, log_dict = out

        results_dict = {'logits': logits}
        if not logits_only:
            results_dict.update({'slide_feats': M})

        if loss_fn is not None and label is not None:
            cls_loss = loss_fn(logits, label)
            loss = cls_loss
            log_dict.update({
                'cls_loss': cls_loss.item(),
                'loss': loss.item()})
            results_dict.update({'loss': loss})
        return results_dict, log_dict
```

**Comment by yang-ze-kang** [comment]
You should also set the 'hidden_size' with 1024.
